src/graphql/index.js

```diff
@@ -14,8 +15,45 @@ export const schema = makeExecutableSchema({
       ] = require(`./resolvers/${fileName}`).default;
       return resolvers;
     }, {}),
+  schemaDirectives: fs
+    .readdirSync(path.join(__dirname, 'directives'))
+    .reduce((directives, fileName) => {
+      directives[
+        fileName.replace(/\.js$/, '')
+      ] = require(`./directives/${fileName}`).default;
+      return directives;
+    }, {}),
 });
 
-const server = new ApolloServer({ schema });
+/**
+ * @param {{ctx: Koa.Context}}
+ * @returns {object}
+ */
+export async function getContext({ ctx: { req } }) {
+  const [userId, nonce] = Buffer.from(
+    (req.headers.authorization || '').replace(/^basic /, ''),
+    'base64'
+  )
+    .toString()
+    .split(':');
+
+  let userContext = null;
+  if (userId && nonce) {
+    const context = await redis.get(userId);
+    if (context && context.nonce === nonce) {
```

Comment:
Where do we generate `nonce`?

Reply:
It's not included in this PR, plan to revisit this logic when implementing LIFF.

Actually I think I will use JWT (introduced in #172 ) to replace nonce, since
1. JWT includes a signature so it cannot be forged
2. We can mix data into JWT for client-side use, such as expire timestamp

Reply:
Actually I found that the recommended way for LINE to make sure a request come from LIFF is via [token validation](https://developers.line.biz/en/docs/line-login/take-over-session/#transfer-login-session):

1. LIFF gets access token via [`getAccessToken` API](https://developers.line.biz/en/reference/liff/#get-access-token)
2. When LIFF invokes server API, attach the access token
3. API server verifies if the token is valid with LINE via [social API](https://developers.line.biz/en/reference/social-api/#verify-access-token) `/oauth2/v2.1/verify?access_token=`

Ref: https://www.facebook.com/groups/linebot/permalink/2489421594721745/

In this way, server do not need to send any form of `nonce` / JWT to LIFF using URL params. Instead, it is LINE server that generates a token (to LIFF) and is asked to verify (from API server).

Reply:
Note:

In the end I decided to use JWT, implemented in this PR: https://github.com/cofacts/rumors-line-bot/pull/178/files

The reason why I use JWT is that I need to preserve state in the JWT, it's not something that LINE server knows, thus access tokens from LINE does not meet mu need.